src/Cardano/Wallet/ChainProducer/RustCardano.hs

```diff
@@ -0,0 +1,139 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Cardano.Wallet.ChainProducer.RustCardano
+    ( RustBackend
+    , runRustBackend
+    ) where
+
+import Control.Exception
+    ( Exception, tryJust )
+import Control.Monad.Except
+import Control.Monad.Reader
+import Prelude
+
+import Cardano.Wallet.ChainProducer
+import Cardano.Wallet.ChainProducer.RustCardano.NetworkLayer
+    ( NetworkLayer (..), NetworkLayerError )
+import Cardano.Wallet.Primitive
+    ( Block (..), BlockHeader (..) )
+import Cardano.Wallet.Slotting
+    ( EpochIndex, SlotCount, SlotId (..), addSlots, slotNext, slotsPerEpoch )
+
+blockSlot :: BlockHeader -> SlotId
+blockSlot bh = SlotId (epochIndex bh) (slotNumber bh)
+
+-- | Calculates which epochs to fetch, given a quantity of blocks, and the start
+-- point.  Takes into account the latest slot available, and that the most
+-- recent epoch is not available in a pack file.
+epochRange :: SlotCount -- ^ Number of blocks
+           -> SlotId -- ^ Start point
+           -> SlotId -- ^ Latest slot available
+           -> [EpochIndex]
+epochRange numBlocks (SlotId startEpoch startSlot) (SlotId tipEpoch _)
```

Comment:
One issue I can think of:

Although we know how many _slots_ are in an epoch, we don't know how many _blocks_ are contained in a particular epoch (because any given slot may be missing its block). Or in other words, blocks and slots are not in 100% correspondence.

Isn't it therefore possible for this function to underestimate the range of epochs required?

Perhaps if the function were to state its assumption, that it only counts based on slots, and not blocks, then it would be clearer.

Reply:
Yes you're right. It assumes slots not blocks.